openedx-unsupported/edx-analytics-dashboard
# Adds a "Download CSV" button to the Learner List View

Adds a button to the Course Learners page, which allows the selected learners list to be downloaded in CSV format.

This report is proxied as a plain text response from the Learner Analytics Data API.

This feature is disabled by default, and can be enabled using the `enable_learner_download` waffle switch.

**Discussions**: Architecture discussed extensively with @mulby and the analytics team, and UI changes designed by Alyssa Boehm.

**Dependencies**: [edx-analytics-data-api #128](https://github.com/edx/edx-analytics-data-api/pull/128)

**Screenshots**:

With no search filters:

![screen shot 2016-08-10 at 5 13 12 pm](https://cloud.githubusercontent.com/assets/7556571/17545776/cc822a2c-5f1d-11e6-9e40-8d82675e5c7a.png)

With search filters:
![screen shot 2016-08-10 at 5 14 15 pm](https://cloud.githubusercontent.com/assets/7556571/17545797/f1ad83a0-5f1d-11e6-8bff-fac0bc3e8657.png)

Responsive positioning on small screen:
![screen shot 2016-08-10 at 5 15 47 pm](https://cloud.githubusercontent.com/assets/7556571/17545831/1e1dcaa8-5f1e-11e6-9813-5b37c7d1cc75.png)

**Partner information**: 3rd party-hosted open edX instance

**Testing instructions**:
1. See testing setup instructions on [edx-analytics-data-api #128](https://github.com/edx/edx-analytics-data-api/pull/128).
2. Run the LMS, Analytics elasticsearch service, and analytics_data_api.  Ensure that you can see your course in the Analytics API http://localhost:8100/docs/#!/api/Learner_List.
3. Update `/edx/etc/insights.xml` to store your Analytics API client token, and your desired CSV field list, e.g.,
   
   ```
   DATA_API_AUTH_TOKEN: your-api-token
   ```
4. Enable the Learner view in Insights:
   
   ```
   sudo su insights
   ./manage.py waffle_flag display_learner_analytics --staff --create
   ./manage.py waffle_switch enable_course_api on --create
   ```
5. Launch Insights, visit the Learners page, and ensure that there is no `Download CSV` button like the one in the above screenshots.
6. Enable this feature switch:
   
   ```
   ./manage.py waffle_switch enable_learner_download on --create 
   ```
7. Relaunch Insights and visit the Learners page to see the new `Download CSV` button.  Click on it to download the CSV file, which should contain all the fields in the learner API call, in alphabetical order.
8. Update `/edx/etc/insights.xml` to update the CSV field list, e.g.,
   
   ```
   LEARNER_API_LIST_DOWNLOAD_FIELDS: username,name,email,cohort,enrollment_date,last_updated
   ```
9. Relaunch Insights, and download the CSV report again.  It should contain only the fields specified in your `settings.LEARNER_API_LIST_DOWNLOAD_FIELDS`, in that order.
10. Load the CSV report URL in an incognito browser, and ensure that you see a Permission Denied error.

**Reviewers**
- [x] @dsjen 
- [ ] @thallada 
- [x] Accessibility - @clrux

**Pre-Merge Checklist**
- [x] Rebase with current master
- [x] Squash into single commit
- [ ] Ensure tests pass
